src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -54,366 +57,1142 @@ internal static class ThreadPoolGlobals
+                    EnqueueSlow(item);
                 }
             }
 
-            public static void Remove(WorkStealingQueue queue)
+            /// <summary>
+            /// Slow path for enqueue, adding a new segment if necessary.
+            /// </summary>
+            private void EnqueueSlow(object item)
             {
-                Debug.Assert(queue != null);
-                while (true)
+                for (; ; )
                 {
-                    WorkStealingQueue[] oldQueues = _queues;
-                    if (oldQueues.Length == 0)
+                    GlobalQueueSegment currentSegment = _enqSegment;
+                    if (currentSegment.TryEnqueue(item))
                     {
                         return;
                     }
 
-                    int pos = Array.IndexOf(oldQueues, queue);
-                    if (pos == -1)
+                    // take the lock to add a new segment
+                    // we can make this optimistically lock free, but it is a rare code path
+                    // and we do not want stampeding enqueuers allocating a lot of new segments when only one will win.
+                    lock (_addSegmentLock)
                     {
-                        Debug.Fail("Should have found the queue");
-                        return;
+                        if (currentSegment == _enqSegment)
+                        {
+                            // Make sure that no more items could be added to the current segment.
+                            // NB: there may be some strugglers still finishing up out-of-order enqueues
+                            //     TryDequeue knows how to deal with that.
+                            currentSegment.EnsureFrozenForEnqueues();
+
+                            // We determine the new segment's length based on the old length.
+                            // In general, we double the size of the segment, to make it less likely
+                            // that we'll need to grow again.  
+                            int nextSize = Math.Min(currentSegment._slots.Length * 2, MaxSegmentLength);
+                            var newEnq = new GlobalQueueSegment(nextSize);
+
+                            // Hook up the new enqueue segment.
+                            currentSegment._nextSegment = newEnq;
+                            _enqSegment = newEnq;
+                        }
                     }
+                }
+            }
 
-                    var newQueues = new WorkStealingQueue[oldQueues.Length - 1];
-                    if (pos == 0)
+            /// <summary>
+            /// Removes an object at the bottom of the queue
+            /// Returns null if the queue is empty.
+            /// </summary>
+            internal object Dequeue()
+            {
+                var currentSegment = _deqSegment;
+                if (!currentSegment.CanSteal)
```

Comment:
It helps a little. It happens relatively often that the global queue is empty.

However we are already not on the fastest path (which would be Pop), so I need to think/measure whether removing the check may benefit bursty case.

---
In reply to: [250367308](https://github.com/dotnet/coreclr/pull/18403#discussion_r250367308) [](ancestors = 250367308)

Reply:
For global work items I guess this would be the fastest path. The case of the queue being empty occurs frequently in bursty cases but I would imagine the case of it not being empty would generally be more frequent.

Reply:
For very short bursts an empty queue may also be very frequent. However, it may also increase the possibility of an incorrect failure to dequeue due to memory ordering. When can you be sure that the global queue is actually empty?

Reply:
I haven't looked at the queue implementations closely enough yet, but `CanSteal` doesn't really make sense for a global queue, would suggest either moving it to the local queue's implementation or renaming it to something like `IsEmpty`.

Reply:
Unlike for global queues, CanSteal on local queues would not be precise enough. You can't tell just by looking at enq/deq ends if there is an unfinished enqueue in progress.

A reliable enough `IsEmpty` for local queue would need to lock the Dequeue slot and find it empty. That is much of the work that Dequeue does, so such `IsEmpty` would not be a cheap easy-out.

---
In reply to: [250442611](https://github.com/dotnet/coreclr/pull/18403#discussion_r250442611) [](ancestors = 250442611)

Reply:
We need to be sure that if a new thread arrives at the same time as we read `numOutstandingThreadRequests` in Enqueue, it will see that we have just made a write.

The incoming thread will see that Enqueue has moved, since global Enqueue is updated via CAS prior reading `numOutstandingThreadRequests` and incoming thread also does CAS - to reduce `numOutstandingThreadRequests`. 

If the thread sees `enq == deq`, it means the items it is responsible for (that were enqueued at the time of reading numOutstandingThreadRequests) are gone. 
That only works for global queue though, since its queue ends are updated via via CAS and monotonously. 

---
In reply to: [250442297](https://github.com/dotnet/coreclr/pull/18403#discussion_r250442297) [](ancestors = 250442297)

Reply:
> A reliable enough IsEmpty for local queue would need to lock the Dequeue slot and find it empty. That is much of the work that Dequeue does, so such IsEmpty would not be a cheap easy-out.

My comment was only about the name, it sounds to me like `CanSteal` for a local queue == `IsEmpty`. How reliable it must be is a separate issue and the same can be said about `CanSteal`.

Reply:
The name was just from the similar helper in the current threadpool.    `IsEmpty` is indeed better. 

---
In reply to: [250474522](https://github.com/dotnet/coreclr/pull/18403#discussion_r250474522) [](ancestors = 250474522)